kafkabp/consumer.go

```diff
@@ -0,0 +1,222 @@
+package kafkabp
+
+import (
+	"context"
+	"io"
+	"sync"
+	"sync/atomic"
+
+	"github.com/Shopify/sarama"
+
+	"github.com/reddit/baseplate.go/metricsbp"
+	"github.com/reddit/baseplate.go/tracing"
+)
+
+// ConsumeMessageFunc is a function type for consuming consumer messages.
+type ConsumeMessageFunc func(ctx context.Context, msg *sarama.ConsumerMessage) error
+
+// ConsumeErrorFunc is a function type for consuming consumer errors.
+type ConsumeErrorFunc func(err error)
+
+// consumer is an instance of a Kafka consumer.
+type consumer struct {
+	cfg ConsumerConfig
+	sc  *sarama.Config
+
+	consumer           atomic.Value // sarama.Consumer
+	partitions         atomic.Value // []int32
+	partitionConsumers atomic.Value // []sarama.PartitionConsumer
+
+	closed          int64
+	consumeReturned int64
+	offset          int64
+
+	wg sync.WaitGroup
+}
+
+// Consumer defines the interface of a consumer struct.
+type Consumer interface {
+	io.Closer
+
+	Consume(ConsumeMessageFunc, ConsumeErrorFunc) error
+
+	// IsHealthy returns false after Consume returns.
+	IsHealthy() bool
+}
+
+// NewConsumer creates a new Kafka consumer. Unlike a group consumer (which
+// delivers every message exactly once by having one ClientID assigned to every
+// consumer in the group), this consumer is used for consuming some
+// configuration or data by all running consumer instances. This is why the
+// ClientID provided to NewConsumer's ConsumerConfig must be unique.
+func NewConsumer(cfg ConsumerConfig) (Consumer, error) {
+	sc, err := cfg.NewSaramaConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	kc := &consumer{
+		cfg:    cfg,
+		sc:     sc,
+		offset: sc.Consumer.Offsets.Initial,
+	}
+
+	// Initialize Sarama consumer and set atomic values.
+	if err := kc.reset(); err != nil {
+		return nil, err
+	}
+
+	return kc, nil
+}
+
+func (kc *consumer) getConsumer() sarama.Consumer {
+	c, _ := kc.consumer.Load().(sarama.Consumer)
+	return c
+}
+
+func (kc *consumer) getPartitions() []int32 {
+	p, _ := kc.partitions.Load().([]int32)
+	return p
+}
+
+func (kc *consumer) getPartitionConsumers() []sarama.PartitionConsumer {
+	pc, _ := kc.partitionConsumers.Load().([]sarama.PartitionConsumer)
+	return pc
+}
+
+// reset recreates the consumer and assigns partitions.
+func (kc *consumer) reset() error {
+	if c := kc.getConsumer(); c != nil {
+		if err := c.Close(); err != nil {
+			kc.cfg.Logger.Log(context.Background(), "kafkabp.consumer.reset: Error closing the consumer:"+err.Error())
+		}
+	}
+
+	rebalance := func() error {
+		c, err := sarama.NewConsumer(kc.cfg.Brokers, kc.sc)
+		if err != nil {
+			return err
+		}
+
+		partitions, err := c.Partitions(kc.cfg.Topic)
+		if err != nil {
+			return err
+		}
+
+		kc.consumer.Store(c)
+		kc.partitions.Store(partitions)
+		return nil
+	}
+
+	err := rebalance()
+	if err != nil {
+		metricsbp.M.Counter("kafka.consumer.rebalance.failure").Add(1)
```

Comment:
:no_bell: not sure if pointed out elsewhere but I assume we want metric on/off configurable in the future?